app/models/sources/linode.rb

```diff
@@ -0,0 +1,190 @@
+module Sources
+  class Linode < Sources::Fog
+    class << self
+      def required_fields
+        super.merge({
+          api_key: :password,
+          root_password: :password,
+          image_name: :string,
+          flavor_name: :string,
+          kernel_name: :string,
+          data_center_name: :string
+        })
+      end
+    end
+
+    ID_TYPES = ['image', 'flavor', 'kernel', 'data_center']
+
+    # Connect to Linode to make sure our config is valid
+    def validate_config!
+      valid = false
+      begin
+        names_to_ids # Force re-evaluation of ids
+        valid = (!image_id.nil? && !flavor_id.nil? && !kernel_id.nil? && !data_center_id.nil?)
+      rescue Exception => e
+        add_error(e.message)
+      end
+      valid
+    end
+
+    private
+
+    # server_is_proxy_type?(server)
+    # given a server, determine if it is running a proxy
+    def server_is_proxy_type?(server)
+      # Linode/Fog does not give access to flavor/image/etc data for a server
+      # Parsing it from the proxy name instead
+      server_details = parse_proxy_name(server.name)
+      return server_details[:image_id] == self.image_id \
+          && server_details[:flavor_id] == self.flavor_id \
+          && server_details[:kernel_id] == self.kernel_id \
+          && server_details[:data_center_id] == self.data_center_id if !server_details.nil?
+    end
+
+    def connection
+      @connection ||= ::Fog::Compute.new(
+        :provider => 'Linode',
+        :linode_api_key => config['api_key']
+      )
+    end
+
+    def server_by_proxy(proxy)
+      return NoServer unless validate_config!
+      server = connection.servers.select do |s|
+        s.public_ip_address == proxy.host
+      end.first
+
+      server || NoServer
+    end
+
+    # The ID_TYPES methods all are very similar, so dynamically create all
+    # of them at once
+    ID_TYPES.each do |id_type|
+      class_eval <<-RUBY
+        # image_id()
+        # flavor_id()
+        # kernel_id()
+        # data_center_id()
+        # Get the image/flavor/kernel/data center id.
+        # If the id is undefined, translate image/flavor/data center name to id by
+        # connecting to Linode.
+        # Returns cached value if already calculated
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the image/flavor/kernel/data center on Linode
+        #   - nil if not found
+        def #{id_type}_id
+          key = '#{id_type}_id'
+          return config[key] if config.has_key?(key)
+          names_to_ids
+          config[key]
+        end
+
+        # retrieve_image_id()
+        # If the id is not found then adds to the source's persistent error log
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the image on Linode
+        #   - nil if not found
+        def retrieve_image_id
```

Comment:
This method is being defined 5 times, 1 time for each loop of `ID_TYPES.each` and an additional time when `id_type == 'image'`.  If `'image'` were the last item in the `ID_TYPES` array then the incorrect version would supersede the version defined here.  You need a separate loop to define the `retrieve_#{id_type}_id` methods that does not include `'image'` as an id_type.